\documentclass{beamer}
\usepackage{BioconductorSlides}

\title{An Overview of Genomic Data Analysis in \Bioconductor}
\author{Martin Morgan\\ Roswell Park Cancer Institute \\ Buffalo, NY, USA}
\date{9 September 2015}

\begin{document}

\maketitle

\begin{frame}{Abstract}
  \Bioconductor{} is a collection of more than 1,000 individually
  code-reviewed software package, hundreds more annotation and
  experiment data packages, and specialized data structures for
  various domains. This talk will provide a high-level overview of
  \Bioconductor{} in the domains of gene expression, DNA variant
  calling, flow cytometry, proteomics and metabolomics. This talk will
  provide users with an up-to-date overview of \Bioconductor's
  offerings and recent developments in these diverse domains. It will
  also review options available to users wishing to analyze public
  data from sources such as The Cancer Genome Atlas, the Gene
  Expression Omnibus, and ArrayExpress, and data distributed by
  \Bioconductor{} itself.
\end{frame}

\section*{Introduction}

\begin{frame}{\R{} and \Bioconductor{}}
  \R{}
  \begin{itemize}
  \item \url{http://r-project.org}
  \item Open-source statistical programming language; widely used in
    academia, finance, pharma, \ldots
  \item Core language, `base' and $>5000$ contributed packages
  \item Interactive sessions, scripts, packages
  \end{itemize}
  \Bioconductor{}
  \begin{itemize}
  \item \url{http://bioconductor.org}
  \item Analysis and comprehension of high-throughput genomic data
  \item Themes: rigorous statistical analysis; reproducible work
    flows; integrative analysis
  \item $13$ years old, $1024$ software packages
  \item Widely used and well-respected
  \end{itemize}
\end{frame}

\section*{Principles}

\begin{frame}[t]{Why use \R{} / \Bioconductor{}?}
  \bigskip\par
  Hallmarks of effective computational software
  \begin{enumerate}
  \item Extensive: \only<2-5>{software, annotation, integration}
    \only<1>{\textbf{\Bioconductor{}} has\ldots
      \begin{itemize}
      \item Software packages for diverse analysis tasks
      \item Annotation (gene and genome) resources to place
        statistical analysis in biological context
      \item Integration across work flows, statistical analysis tasks,
        \ldots
      \end{itemize}
    }
  \item Statistical: \only<1, 3-5>{volume, technology, experimental design}
    \only<2>{
      \begin{itemize}
      \item Volume of data requires statistical treatment
      \item Technology introduces statistical artifacts that need to
        be accommodated
      \item Designed experiments need statistical treatment, often
        involve small sample sizes (even though the data is big!), and
        require appropriate analysis
      \item \textbf{\R{} / \Bioconductor{}} a `natural' for
        statistical analysis
      \end{itemize}
    }
  \item Reproducible: \only<1-2, 4-5>{long-term, multi-participant science}
    \only<3>{
      \begin{itemize}
      \item Long-term analysis projects (days, months, years) need to
        be performed in a way that is reproducible
      \item Multi-participant science requires facilities that bridge,
        e.g., statistical analysis and communication with biologist
        collaborators.
      \item Many \textbf{features of \Bioconductor{}} encourage
        reproducible research, e.g., script, vignettes, packages,
        \ldots
      \end{itemize}
    }
  \item Leading edge: \only<1-3,5>{novel, technology-driven}
    \only<4>{
      \begin{itemize}
      \item Science research questions changes rapidly, so
        \emph{usable} software needs to keep pace without necessarily
        being \emph{finished}
      \item Software needs to keep pace with rapidly changing
        technology
      \item \textbf{\Bioconductor{} package structure and user
          community} enable project to closely tracks leading edge
        science
      \end{itemize}
    }
  \item Accessible: \only<1-4>{affordable, transparent, usable}
    \only<5>{
      \begin{itemize}
      \item Research software needs to be affordable, both in terms of
        money and time
      \item Science requires transparent algorithms, so that merits
        can be evaluated, problems identified and corrected
      \item Software has to be usable in the sense that it
        \emph{works}, and is \emph{documented}
      \item \textbf{\Bioconductor{}} is free and open, with extensive
        documentation and an active and supportive user community
      \end{itemize}
    }
  \end{enumerate}
\end{frame}

\begin{frame}{Major scientific domains}
  Sequencing
  \begin{itemize}
  \item Differential expression
  \item ChIP-seq, methyl-seq
  \item Called variants
  \item Copy number, structure
  \item Microbiome
  \end{itemize}
  Methylation, expression, and other microarrays
  \smallskip\par
  Flow cytometry
  \smallskip\par
  Proteomics
  \smallskip\par
  \ldots
\end{frame}

\begin{frame}{Where does \Bioconductor{} fit in? E.g., sequencing}
  \centering{\includegraphics[width=\textwidth]{our_figures/SequencingEcosystem.png}}
\end{frame}

\section*{Exemplars}

\begin{frame}{Exemplar: Batch Effects}
  Leek et al., 2010, Nature Reviews Genetics 11,
  \href{http://www.nature.com/nrg/journal/v11/n10/abs/nrg2825.html}{733-739},
  Leek \& Story
  \href{https://doi.org/10.1371/journal.pgen.0030161}{PLoS Genet
    3(9): e161}
  \begin{columns}
    \column{.5\textwidth} 
    \begin{itemize}
    \item Scientific finding: pervasive batch effects
    \item Statistical insights: surrogate variable analysis: identify
      and build surrogate variables; remove known batch effects
    \item Benefits: reduce dependence, stabilize error rate estimates,
      and improve reproducibility
    \end{itemize}
    \Bioconductor{} support: \Biocpkg{sva}; also \Biocpkg{RUVSeq}
    \column{.5\textwidth}
    \only<1>{
      \includegraphics[width=\textwidth]{T1_figures/nrg2825-f2.jpg}
      \par{\small HapMap samples from one facility, ordered by
        date of processing.  From }
    }
    \only<2>{
      \begin{enumerate}
      \item Remove signal due to variable(s) of interest
      \item Identify subset of genes driving orthogonal signatures
        of EH
      \item Build a surrogate variable based on full EH signature of
        that subset
      \item Include significant surrogate variables as covariates
      \end{enumerate}
      EH: expression heterogeneity
    }
  \end{columns}
\end{frame}

\begin{frame}{Exemplar: Epigenetics}
  Hansen et al., 2011, Nature Genetics 43, 
  \href{http://www.nature.com/ng/journal/v43/n8/full/ng.865.html}{768-775}
  \begin{itemize}
  \item Scientific finding: stochastic methylation variation of
    cancer-specific de-methylated regions (DMR), distinguishing cancer from
    normal tissue, in several cancers.
  \item Statistical challenges: smoothing, non-specific filtering, $t$
    statistics, find DMRs
  \end{itemize}
  \bigskip\par
  \includegraphics[width=\textwidth]{T1_figures/bsseq_analysis-1.png}
  \medskip\par \Bioconductor{} support: whole-genome (\Biocpkg{bsseq})
  or reduced representation (\Biocpkg{MethylSeekR}) bisulfite
  sequencing; Illumina 450k arrays (\Biocpkg{minfi}) -- Peter Hickey's
  talks(!)
\end{frame}

\begin{frame}{Exemplar: Differential Representation}
  Haglund et al., 2012
  \href{http://www.ncbi.nlm.nih.gov/pubmed/23024189}{J Clin Endocrin
    Metab}
  \bigskip\par
  \begin{columns}
    \column{.5\textwidth}
    \begin{itemize}
    \item Scientific finding: identify genes whose expression is
      regulated by estrogen receptors in parathyroid adenoma cells
    \item Statistical challenges: between-sample normalization;
      appropriate statistical model; efficient estimation; \ldots
    \end{itemize}
    \column{.5\textwidth}
    \includegraphics[width=\textwidth]{T1_figures/DESeq2_parathyroid-plotMApadjchange.png}
  \end{columns}
  \bigskip\par\Bioconductor{} support: \Biocpkg{DESeq2},
  \Biocpkg{edgeR}, many statistical `lessons learned' from
  microarrays; extensive integration with down-stream
  tools. \Biocpkg{limma} \Rfunction{voom} when transformation to an
  approximate normal distribution is appropriate.
\end{frame}

\begin{frame}{Exemplar: Visualization}
  \begin{columns}
    \column{.5\textwidth}
    \Biocpkg{Gviz}\par
    \only<1-2>{
    \begin{itemize}
    \item Track-like visualizations
    \item Data panels
    \item Fully integrated with \Bioconductor{} sequence
      representations
    \end{itemize}
    }
    \Biocpkg{ggbio}\par
    \only<3>{
    \begin{itemize}
    \item Comprehensive visualizations
    \item \Rfunction{autoplot} file and data types
    \item Fully integrated with \Bioconductor{} sequence
      representations
    \end{itemize}
    }
    \Biocpkg{epivizr}\par
    \only<4>{
      \begin{itemize}
      \item Genome browser with socket communication to \R{}
      \item Fully integrated with \Bioconductor{} sequence
        representations
      \end{itemize}
    }
    \column{.5\textwidth}
    \only<1>{\includegraphics[width=\textwidth]{T1_figures/Gviz-vignette-1.png}}
    \only<2>{\includegraphics[width=\textwidth]{T1_figures/Gviz-vignette-2.png}}
    \only<3>{\includegraphics[width=\textwidth]{T1_figures/ggbio-vignette-1.png}}
    \only<4>{\includegraphics[width=\textwidth]{T1_figures/epivisr.png}}
  \end{columns}
\end{frame}

\begin{frame}[fragile]{Exemplar: Three Types of Annotation Resources}
  \begin{columns}
    \column{.5\textwidth}
    \begin{itemize}
    \item Gene models (\Rpackage{TxDb.*}), genome sequences
      (\Rpackage{BSgenome.*}), and identifier maps (\Rpackage{org.*}).
    \item Web-based annotations (e.g., \Biocpkg{biomaRt},
      \Biocpkg{pathway}, \Biocpkg{PSICQUIC}) and experiment data
      repositories (e.g., \Biocpkg{GEOquery}, \Biocpkg{SRAdb}).
    \item \Biocpkg{AnnotationHub} to access consortium and other
      genome-scale data collections.
    \end{itemize}
    \column{.5\textwidth}
\begin{verbatim}
library(rtracklayer)
library(AnnotationHub)
hub <- AnnotationHub()

## RoadMap E126 H3K4ME2 marks
E126 <- hub[["AH29817"]]
## UCSC hg19 -> hg38 liftOver
chain <- hub[["AH14150"]]

E126hg38 <- 
    liftOver(E126, chain)
\end{verbatim}
  \end{columns}
\end{frame}

\begin{frame}{Exemplars: \emph{etcetera!}}
  \begin{columns}[t]
    \column{.5\textwidth}
    ChIP-seq
    \begin{itemize}
    \item \Biocpkg{csaw} -- window-based differential binding.
    \end{itemize}
    Variants
    \begin{itemize}
    \item \Biocpkg{VariantFiltering} -- characterization of variants.
    \item \Biocpkg{SomaticSignatures} -- mutational signatures
      associated with SNPs.
    \end{itemize}
    Sequencing work flows / reports
    \begin{itemize}
    \item \Biocpkg{QuasR}, \Biocpkg{systemPipeR}.
    \end{itemize}
    \column{.5\textwidth}
    Pathways
    \begin{itemize}
    \item \Biocpkg{pathview} integration and visualization.
    \end{itemize}
    Flow cytometry
    \begin{itemize}
    \item \Biocpkg{flowCore}; 
      `\href{http://bioconductor.org/packages/release/BiocViews.html\#___FlowCytometry}{FlowCytometry}'
      biocViews term.
    \end{itemize}
    Mass Spec / Proteomics
    \begin{itemize}
    \item \Biocpkg{mzR}; proteomics
      \href{http://bioconductor.org/help/workflows/proteomics/}{workflow}.
    \end{itemize}
    Visualization
    \begin{itemize}
    \item \Biocpkg{ComplexHeatmap}, \Biocpkg{HilbertCurve}
    \end{itemize}
  \end{columns}
\end{frame}

\begin{frame}{Summary}
  More information
  \begin{itemize}
  \item \url{http://bioconductor.org}
  \item \url{https://support.bioconductor.org}
  \end{itemize}
  Why \R{} / \Bioconductor?
  \begin{itemize}
  \item Open-source statistical programming language
  \item Open-development packages!
  \item Extensive, statistical, reproducible, leading edge, accessible.
  \item Sequencing, methylation and other arrays, flow cytometery,
    proteomics, ...
  \end{itemize}
  Increasingly easy to learn and use
  \begin{itemize}
  \item \href{http://rstudio.org}{RStudio}, Coursera, `shiny'
    applications\ldots
  \end{itemize}
\end{frame}

\begin{frame}{Acknowledgments}
  
  \medskip\par
  Core (Seattle / Buffalo): Jim Hester, Jim Java, Brian Long, Valerie
  Obenchain, Herv\'e Pag\`es, Dan Tenenbaum. Recent alumni: Marc
  Carlson, Sonali Arora, Nate Hayden, Paul Shannon.
  \medskip\par
  International developer community, technical and scientific advisory
  boards.
  \medskip\par
  The research reported in this presentation was supported by the
  National Cancer Institute and the National Human Genome Research
  Institute of the National Institutes of Health under award numbers
  U41HG004059 and U24CA180996, and the National Science Foundation
  under Award number 1247813. The content is solely the responsibility
  of the authors and does not necessarily represent the official views
  of the National Institutes of Health or the National Science
  Foundation.

\end{frame}

\end{document}
